File: sorting/quick-sort/javascript/quick-sort.js
/**
 * Sorts an array using the Quick Sort algorithm (Lomuto partition scheme).
 *
 * @param {number[]} arr - The array to be sorted.
 * @returns {number[]} A new sorted array (original is not modified).
 *
 * Time Complexity:
 * - Best/Average: O(n log n)
 * - Worst: O(n²) — rare with random data
 * Space Complexity: O(log n) due to recursion stack
 */
function quickSort(arr) {
    // Create a copy to avoid mutating the original array
    const sorted = [...arr];

    // Helper function to sort the array in-place using recursion
    function sort(low, high) {
        if (low < high) {
            // Partition the array and get the pivot index
            const pivotIndex = partition(low, high);
            // Recursively sort left and right partitions
            sort(low, pivotIndex - 1);
            sort(pivotIndex + 1, high);
        }
    }

    // Lomuto partition scheme: uses last element as pivot
    function partition(low, high) {
        const pivot = sorted[high]; // Choose last element as pivot
        let i = low - 1; // Index of smaller element

        for (let j = low; j < high; j++) {
            if (sorted[j] <= pivot) {
                i++;
                // Swap elements
                [sorted[i], sorted[j]] = [sorted[j], sorted[i]];
            }
        }

        // Place pivot in correct position
        [sorted[i + 1], sorted[high]] = [sorted[high], sorted[i + 1]];
        return i + 1; // Return pivot index
    }

    // Start sorting from full array range
    if (sorted.length > 1) {
        sort(0, sorted.length - 1);
    }

    return sorted;
}

// Example Usage
const numbers = [10, 80, 30, 90, 40, 50, 70];
console.log("Original Array: ", numbers);
console.log("Sorted Array:   ", quickSort(numbers));

// Test with edge cases
console.log("Empty Array:    ", quickSort([]));
console.log("Single Element: ", quickSort([42]));
console.log("Already Sorted: ", quickSort([1, 2, 3, 4, 5]));
console.log("Reverse Sorted: ", quickSort([5, 4, 3, 2, 1]));

File: sorting/quick-sort/csharp/QuickSort.cs
using System;

/**
 * Sorts an array using the Quick Sort algorithm with Lomuto partition scheme.
 */
public class QuickSort
{
    /**
     * Sorts an array using Quick Sort (Lomuto partition).
     *
     * @param arr The input array to be sorted.
     * @return A new sorted array (original is not modified).
     *
     * Time Complexity:
     * - Best/Average: O(n log n)
     * - Worst: O(n²) — rare with random data
     * Space Complexity: O(log n) due to recursion stack
     */
    public static int[] QuickSortArray(int[] arr)
    {
        // Handle null or small arrays
        if (arr == null || arr.Length <= 1)
            return (int[])arr?.Clone() ?? new int[0];

        // Create a copy to avoid mutating the original array
        int[] sorted = (int[])arr.Clone();

        // Sort the entire array
        Sort(sorted, 0, sorted.Length - 1);

        return sorted;
    }

    /**
     * Recursively sorts the subarray from low to high.
     */
    private static void Sort(int[] arr, int low, int high)
    {
        if (low < high)
        {
            // Partition the array and get pivot index
            int pivotIndex = Partition(arr, low, high);
            // Recursively sort left and right partitions
            Sort(arr, low, pivotIndex - 1);
            Sort(arr, pivotIndex + 1, high);
        }
    }

    /**
     * Partitions the array using Lomuto scheme (last element as pivot).
     *
     * @return The final index of the pivot element
     */
    private static int Partition(int[] arr, int low, int high)
    {
        int pivot = arr[high]; // Choose last element as pivot
        int i = low - 1;       // Index of smaller element

        for (int j = low; j < high; j++)
        {
            if (arr[j] <= pivot)
            {
                i++;
                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Place pivot in correct position
        int pivotTemp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = pivotTemp;

        return i + 1; // Return pivot index
    }

    // Example Usage
    static void Main(string[] args)
    {
        int[] numbers = {10, 80, 30, 90, 40, 50, 70};

        Console.WriteLine("Original Array: " + FormatArray(numbers));
        Console.WriteLine("Sorted Array:   " + FormatArray(QuickSortArray(numbers)));

        // Test edge cases
        Console.WriteLine("Empty Array:    " + FormatArray(QuickSortArray(new int[] {})));
        Console.WriteLine("Single Element: " + FormatArray(QuickSortArray(new int[] {42})));
        Console.WriteLine("Already Sorted: " + FormatArray(QuickSortArray(new int[] {1, 2, 3, 4, 5})));
        Console.WriteLine("Reverse Sorted: " + FormatArray(QuickSortArray(new int[] {5, 4, 3, 2, 1})));
    }

    // Helper method to format array output
    static string FormatArray(int[] arr)
    {
        return "[" + string.Join(", ", arr) + "]";
    }
}